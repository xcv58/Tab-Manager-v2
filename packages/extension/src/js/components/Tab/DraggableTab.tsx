import React, { useEffect } from 'react'
import classNames from 'classnames'
import { observer } from 'mobx-react-lite'
import { useDrag } from 'react-dnd'
import DroppableTab from './DroppableTab'
import { ItemTypes } from 'libs/react-dnd'
import { useStore } from 'components/hooks/useStore'
import { getEmptyImage } from 'react-dnd-html5-backend'
import { TabProps } from '../types'

export default observer((props: TabProps) => {
  const { tab } = props
  const { dragStore } = useStore()
  const [dragProps, drag, connectDragPreview] = useDrag({
    item: { type: ItemTypes.TAB },
    canDrag: true,
    begin: () => {
      dragStore.dragStart(tab)
    },
    end: () => {
      dragStore.dragEnd()
    },
    isDragging: () => tab.isSelected,
    collect: (monitor) => {
      return {
        isDragging: monitor.isDragging(),
      }
    },
  })
  const { isDragging } = dragProps
  useEffect(() => {
    connectDragPreview(getEmptyImage())
  })
  return (
    <div ref={drag} className={classNames({ 'opacity-25': isDragging })}>
      <DroppableTab {...props} />
    </div>
  )
})
