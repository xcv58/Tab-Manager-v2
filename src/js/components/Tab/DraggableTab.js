import React from 'react'
import { inject, observer } from 'mobx-react'
import { DragSource, DropTarget } from 'react-dnd'
import Tab from './Tab'
import { ItemTypes } from 'libs'
import Preview from 'components/Preview'

const tabSource = {
  beginDrag (props, monitor, component) {
    const {
      tab,
      dragStore: { dragStart }
    } = props
    dragStart(tab)
    return {}
  },
  endDrag (props, monitor, component) {
    props.dragStore.dragEnd()
  },
  isDragging (props, monitor) {
    return props.tab.isSelected
  }
}

const tabTarget = {
  drop (props) {
    const {
      tab,
      dragStore: { drop }
    } = props
    drop(tab)
  }
}

@inject('dragStore')
@DropTarget(ItemTypes.TAB, tabTarget, (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isOver: monitor.isOver({ shallow: true })
}))
@DragSource(ItemTypes.TAB, tabSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  connectDragPreview: connect.dragPreview(),
  isDragging: monitor.isDragging()
}))
@observer
export default class DraggableTab extends React.Component {
  componentDidMount () {
    const { connectDragPreview } = this.props
    connectDragPreview(document.getElementById('dragPreview'))
  }

  render () {
    const {
      tab: { showTab },
      connectDragSource,
      connectDropTarget,
      isDragging,
      isOver
    } = this.props
    const style = {}
    if (isDragging) {
      style.display = 'none'
    }
    const preview = isOver && <Preview />
    return connectDropTarget(
      connectDragSource(
        <div
          style={{
            ...style,
            ...this.props.style
          }}
        >
          {preview}
          {showTab && <Tab {...this.props} />}
        </div>
      )
    )
  }
}
