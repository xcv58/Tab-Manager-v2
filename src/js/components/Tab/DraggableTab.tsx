import React, { useEffect } from 'react'
import { observer } from 'mobx-react'
import { useDrag, useDrop } from 'react-dnd'
import Tab from './Tab'
import { ItemTypes } from 'libs/react-dnd'
import Preview from 'components/Preview'
import { withTheme } from '@material-ui/core/styles'
import { useStore } from 'components/StoreContext'
import { getEmptyImage } from 'react-dnd-html5-backend'

const DraggableTab = observer(props => {
  const { tab, theme } = props
  const { showTab } = tab
  const { dragStore } = useStore()
  const [dragProps, drag, connectDragPreview] = useDrag({
    item: { type: ItemTypes.TAB },
    canDrag: true,
    begin: () => {
      dragStore.dragStart(tab)
    },
    end: () => {
      dragStore.dragEnd()
    },
    isDragging: () => tab.isSelected,
    collect: monitor => {
      return {
        isDragging: monitor.isDragging()
      }
    }
  })
  const [dropProps, drop] = useDrop({
    accept: ItemTypes.TAB,
    drop: () => {
      dragStore.drop(tab)
    },
    canDrop: () => tab.win.canDrop,
    collect: monitor => {
      return {
        canDrop: monitor.canDrop(),
        isOver: monitor.isOver({ shallow: true })
      }
    }
  })
  const { isDragging } = dragProps
  const { isOver, canDrop } = dropProps
  const style = {}
  if (isDragging) {
    style.display = 'none'
  }
  const tabStyle = {}
  if (isOver && !canDrop) {
    tabStyle.backgroundColor = theme.palette.error.light
  }
  const preview = canDrop && isOver && <Preview />
  useEffect(() => {
    connectDragPreview(getEmptyImage())
  })
  return (
    <div
      ref={node => {
        drag(node)
        drop(node)
      }}
      style={{ ...style, ...props.style }}
    >
      {preview}
      {showTab && <Tab {...props} style={tabStyle} />}
    </div>
  )
})

export default withTheme(DraggableTab)
