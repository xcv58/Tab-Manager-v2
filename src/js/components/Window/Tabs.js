import React from 'react'
import { inject, observer } from 'mobx-react'
import FlipMove from 'react-flip-move'
import DraggableTab from 'components/Tab/DraggableTab'

@inject('windowStore')
@observer
export default class Tabs extends React.Component {
  componentDidMount () {
    window.requestAnimationFrame(this.props.windowStore.windowMounted)
  }

  render () {
    const {
      win: { tabs },
      getScrollbars,
      dragPreview
    } = this.props
    const tabsView = tabs.map(tab => (
      <DraggableTab
        key={tab.id}
        tab={tab}
        {...{ getScrollbars, dragPreview }}
      />
    ))
    return (
      <FlipMove
        duration={255}
        easing='ease-in-out'
        appearAnimation='accordionHorizontal'
        enterAnimation='accordionHorizontal'
        leaveAnimation='accordionHorizontal'
      >
        {tabsView}
      </FlipMove>
    )
  }
}
