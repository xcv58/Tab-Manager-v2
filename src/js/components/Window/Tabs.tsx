import React, { useEffect } from 'react'
import { observer } from 'mobx-react'
import DraggableTab from 'components/Tab/DraggableTab'
import { useStore } from 'components/StoreContext'

export default observer((props) => {
  const { windowStore } = useStore()
  useEffect(() => {
    window.requestAnimationFrame(windowStore.windowMounted)
  }, [])

  const { win } = props
  if (win.hide) {
    return null
  }
  return win.tabs
    .filter((x) => x.isVisible)
    .map((tab) => <DraggableTab key={tab.id} tab={tab} />)
})
