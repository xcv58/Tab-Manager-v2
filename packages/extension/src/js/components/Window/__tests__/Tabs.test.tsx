import React from 'react'
import { connectDropTarget } from 'test'
import { spy, stub } from 'sinon'
import { shallow } from 'enzyme'
import DraggableTab from 'components/Tab/DraggableTab'
import Tabs from 'components/Window/Tabs'

const tabs = [{ id: 1, isVisible: true }, { id: 2, isVisible: true }, { id: 3 }]
const windowMounted = spy()
const props = {
  connectDropTarget,
  dragStore: {
    drop: spy(),
  },
  windowStore: {
    windowMounted,
  },
  win: {
    tabs,
    showTabs: true,
  },
  dragPreview: 'preview node',
}

describe('Tabs', () => {
  it('render correct components', () => {
    const el = shallow(<Tabs {...props} />)
    expect(el.find(DraggableTab).length).toBe(2)
  })

  it('does not render tab with isVisible = `false`', () => {
    const el = shallow(
      <Tabs
        {...props}
        win={{
          ...props.win,
          tabs: tabs.map((x) => ({ ...x, isVisible: false })),
        }}
      />
    )
    expect(el.find(DraggableTab).length).toBe(0)
  })

  it.skip('call requestAnimationFrame with windowStore.windowMounted', () => {
    const requestAnimationFrame = stub(window, 'requestAnimationFrame')
    const el = shallow(<Tabs {...props} />)
    // TODO: trigger useEffect, it's blocked by https://github.com/airbnb/enzyme/issues/2086
    el.update()
    expect(requestAnimationFrame.callCount).toBe(1)
    expect(requestAnimationFrame.args[0]).toEqual([windowMounted])
  })
})
