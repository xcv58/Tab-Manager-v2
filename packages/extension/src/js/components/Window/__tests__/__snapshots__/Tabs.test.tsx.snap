// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs does not render tab with isVisible = \`false\` 1`] = `<div />`;

exports[`Tabs render correct components 1`] = `
<div>
  <div
    class=""
    draggable="true"
  >
    <div>
      <div
        class="flex relative items-center opacity-25 hover:bg-blue-300"
        tabindex="-1"
      >
        <div
          class="group"
        >
          <div
            class="group-hover:hidden"
          >
            <button
              aria-label="Toggle select"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium focus:outline-none focus:ring css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <img
                class="w-6 h-6"
              />
            </button>
          </div>
          <div
            class="focus:outline-none focus:ring hidden group-hover:block"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-imrjgg-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                aria-label="Toggle select"
                class="PrivateSwitchBase-input css-j8yymo"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
        </div>
        <button
          class="group flex flex-col justify-center flex-1 h-12 overflow-hidden text-left m-0 rounded-sm text-base"
          data-mui-internal-clone-element="true"
        >
          <div
            class="w-full overflow-hidden truncate"
          />
          <div
            class="w-full overflow-hidden truncate text-xs opacity-75 group-hover:opacity-100 group-hover:text-black"
          />
        </button>
        <button
          class="inline-flex justify-center w-8 h-8 m-2 text-xl text-red-200 bg-transparent rounded-full disabled:opacity-75 hover:text-red-500 hover:bg-red-100 focus:outline-none focus:ring active:bg-red-300 active:text-red-700"
        >
          x
        </button>
      </div>
    </div>
  </div>
  <div
    class=""
    draggable="true"
  >
    <div>
      <div
        class="flex relative items-center opacity-25 hover:bg-blue-300"
        tabindex="-1"
      >
        <div
          class="group"
        >
          <div
            class="group-hover:hidden"
          >
            <button
              aria-label="Toggle select"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium focus:outline-none focus:ring css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <img
                class="w-6 h-6"
              />
            </button>
          </div>
          <div
            class="focus:outline-none focus:ring hidden group-hover:block"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-imrjgg-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                aria-label="Toggle select"
                class="PrivateSwitchBase-input css-j8yymo"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
        </div>
        <button
          class="group flex flex-col justify-center flex-1 h-12 overflow-hidden text-left m-0 rounded-sm text-base"
          data-mui-internal-clone-element="true"
        >
          <div
            class="w-full overflow-hidden truncate"
          />
          <div
            class="w-full overflow-hidden truncate text-xs opacity-75 group-hover:opacity-100 group-hover:text-black"
          />
        </button>
        <button
          class="inline-flex justify-center w-8 h-8 m-2 text-xl text-red-200 bg-transparent rounded-full disabled:opacity-75 hover:text-red-500 hover:bg-red-100 focus:outline-none focus:ring active:bg-red-300 active:text-red-700"
        >
          x
        </button>
      </div>
    </div>
  </div>
</div>
`;
